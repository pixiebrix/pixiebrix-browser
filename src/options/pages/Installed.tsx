import Table from "react-bootstrap/Table";
import isEmpty from "lodash/isEmpty";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import React, { useMemo } from "react";
import extensionPointRegistry from "@/extensionPoints/registry";
import { optionsSlice, OptionsState } from "../slices";
import Button from "react-bootstrap/Button";
import { useToasts } from "react-toast-notifications";
import { PageTitle } from "@/layout/Page";
import { useExtensionPermissions } from "@/permissions";
import { BeatLoader, GridLoader } from "react-spinners";
import Card from "react-bootstrap/Card";
import {
  faCheck,
  faExclamation,
  faCubes,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Link } from "react-router-dom";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import {
  ExtensionValidationResult,
  useExtensionValidator,
} from "@/validators/generic";
import { IExtension, ServiceLocator } from "@/core";
import useExtensionStore from "../extensionStore";
import { bindActionCreators } from "redux";
import LazyLocatorFactory from "@/services/locator";

const { removeExtension } = optionsSlice.actions;

type RemoveAction = ({
  extensionId,
  extensionPointId,
}: {
  extensionId: string;
  extensionPointId: string;
}) => void;

function validationMessage(validation: ExtensionValidationResult) {
  let message = "Invalid Configuration";
  if (validation.notConfigured.length) {
    const services = validation.notConfigured.map((x) => x.serviceId);
    if (services.length > 1) {
      message = `The following services have no configurations: ${services.join(
        ", "
      )}`;
    } else {
      message = `You need to add a configuration for ${services[0]}`;
    }
  } else if (validation.missingConfiguration.length) {
    const services = validation.missingConfiguration.map((x) => x.serviceId);
    message = `
      The following services use configurations that no longer exist: ${services.join(
        ", "
      )}`;
  } else if (validation.multipleAuths.length) {
    const services = validation.multipleAuths.map((x) => x.serviceId);
    message = `Multiple configurations exist for these services, you must select which one to use: ${services.join(
      ", "
    )}`;
  } else {
    console.debug("Validation result", validation);
  }
  return message;
}

const ExtensionRow: React.FunctionComponent<{
  extension: IExtension;
  onRemove: RemoveAction;
  locator: ServiceLocator;
}> = ({ extension, onRemove, locator }) => {
  const { id, label, extensionPointId } = extension;
  const { addToast } = useToasts();
  const [hasPermissions, requestPermissions] = useExtensionPermissions(
    extension
  );
  const [validation] = useExtensionValidator(locator, extension);

  const statusElt = useMemo(() => {
    if (hasPermissions == null || validation == null) {
      return <BeatLoader />;
    } else if (validation && !validation.valid) {
      return (
        <span className="text-danger text-wrap">
          <FontAwesomeIcon icon={faExclamation} />{" "}
          {validationMessage(validation)}
        </span>
      );
    } else if (hasPermissions) {
      return (
        <span>
          <FontAwesomeIcon icon={faCheck} /> Active
        </span>
      );
    } else {
      return (
        <Button variant="info" size="sm" onClick={requestPermissions}>
          Grant Permissions
        </Button>
      );
    }
  }, [hasPermissions, requestPermissions, validation]);

  return (
    <tr>
      <td>{extensionPointRegistry.lookup(extensionPointId).name}</td>
      <td>
        <Link to={`/workshop/extensions/${extension.id}`}>{label ?? id}</Link>
      </td>
      <td className="text-wrap">{statusElt}</td>
      <td>
        <Button
          variant="danger"
          size="sm"
          onClick={() => {
            onRemove({ extensionId: id, extensionPointId });
            addToast(`Removed brick ${label ?? id}`, {
              appearance: "success",
              autoDismiss: true,
            });
          }}
        >
          Uninstall
        </Button>
      </td>
    </tr>
  );
};

const Installed: React.FunctionComponent<{
  extensions: IExtension[];
  onRemove: RemoveAction;
}> = ({ extensions, onRemove }) => {
  const locator = useMemo(() => {
    const locatorFactory = new LazyLocatorFactory();
    return locatorFactory.getLocator();
  }, []);

  return (
    <div>
      <PageTitle icon={faCubes} title="Active Bricks" />

      <Row>
        <Col>
          <div className="pb-4">
            <p>
              Here's a list of bricks you currently have installed. You can find
              more to install in the{" "}
              <Link to={"/marketplace"}>Marketplace</Link>
            </p>
          </div>
        </Col>
      </Row>
      <Row>
        <Col xl={9} lg={10} md={12}>
          <Card>
            <Card.Header>Active Bricks</Card.Header>
            <Table>
              <thead>
                <tr>
                  <th>Location</th>
                  <th>Name</th>
                  <th>Status</th>
                  <th>Uninstall</th>
                </tr>
              </thead>
              <tbody>
                {extensions.map((extension) => (
                  <ExtensionRow
                    key={extension.id}
                    extension={extension}
                    locator={locator}
                    onRemove={onRemove}
                  />
                ))}
                {isEmpty(extensions) && (
                  <tr>
                    <td colSpan={4}>
                      No bricks installed yet. Find some in the{" "}
                      <Link to={"/marketplace"}>Marketplace</Link>
                    </td>
                    <td>&nbsp;</td>
                    <td>&nbsp;</td>
                    <td>&nbsp;</td>
                  </tr>
                )}
              </tbody>
            </Table>
          </Card>
        </Col>
      </Row>
    </div>
  );
};

Installed.propTypes = {
  extensions: PropTypes.array,
};

function selectExtensions(state: { options: OptionsState }) {
  return Object.entries(state.options.extensions).flatMap(
    ([extensionPointId, pointExtensions]) =>
      Object.entries(pointExtensions).map(([extensionId, extension]) => ({
        id: extensionId,
        extensionPointId,
        ...extension,
      }))
  );
}

export const WebInstalledContainer: React.FunctionComponent = () => {
  const [state, dispatch] = useExtensionStore();
  const extensions = useMemo(
    () => (state ? selectExtensions(state) : undefined),
    [state]
  );

  // @ts-ignore: YOLO
  const { onRemove } = useMemo(
    // @ts-ignore: YOLO
    () => bindActionCreators({ onRemove: removeExtension }, dispatch),
    [dispatch]
  );

  if (extensions == null) {
    return <GridLoader />;
  }

  return <Installed extensions={extensions} onRemove={onRemove} />;
};

export default connect(
  (state: { options: OptionsState }) => ({
    extensions: selectExtensions(state),
  }),
  { onRemove: removeExtension }
)(Installed);
