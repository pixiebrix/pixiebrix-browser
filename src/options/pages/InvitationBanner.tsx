/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import Banner from "@/components/banner/Banner";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";
import { useGetInvitationsQuery } from "@/services/api";

// eslint-disable-next-line prefer-destructuring -- It breaks EnvironmentPlugin
const SERVICE_URL = process.env.SERVICE_URL;

const InvitationBanner: React.FunctionComponent = () => {
  const { data: invitations } = useGetInvitationsQuery();
  const invitationsAvailable = invitations?.length > 0;

  if (!invitationsAvailable) {
    return null;
  }

  return (
    <Banner variant="info">
      You have team invitations waiting for you! &nbsp;
      <span role="img" aria-label="sparkles">
        ✨
      </span>
      <span role="img" aria-label="envelope">
        ✉️
      </span>
      &nbsp; You can view them on the{" "}
      <a href={`${SERVICE_URL}/team`} target="_blank" rel="noopener noreferrer">
        &quot;My Team&quot; Page <FontAwesomeIcon icon={faExternalLinkAlt} />
      </a>{" "}
      of the Admin Console.
    </Banner>
  );
};

export default InvitationBanner;
