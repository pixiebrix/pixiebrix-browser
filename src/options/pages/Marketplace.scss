.marketplace-component {
  .card-header {
    background-color: #450163;
    color: white;
  }
}
