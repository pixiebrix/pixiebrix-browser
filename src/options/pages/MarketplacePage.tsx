/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React, { useCallback } from "react";
import { connect } from "react-redux";
import { OptionsState } from "../slices";
import GenericMarketplacePage from "@/pages/marketplace/MarketplacePage";
import { push } from "connected-react-router";
import { RecipeDefinition } from "@/types/definitions";
import { useTitle } from "@/hooks/title";

export interface MarketplaceProps {
  installedRecipes: Set<string>;
  navigate: (url: string) => void;
}

const MarketplacePage: React.FunctionComponent<MarketplaceProps> = ({
  installedRecipes,
  navigate,
}) => {
  useTitle("Marketplace");

  const install = useCallback(
    async (x: RecipeDefinition) => {
      navigate(`marketplace/activate/${encodeURIComponent(x.metadata.id)}`);
    },
    [navigate]
  );

  return (
    <GenericMarketplacePage
      installedRecipes={installedRecipes}
      installRecipe={install}
    />
  );
};

export default connect(
  ({ options }: { options: OptionsState }) => ({
    installedRecipes: new Set(
      Object.values(options.extensions).flatMap((extensionPoint) =>
        Object.values(extensionPoint)
          .map((x) => x._recipeId)
          .filter((x) => x)
      )
    ),
  }),
  { navigate: push }
)(MarketplacePage);
