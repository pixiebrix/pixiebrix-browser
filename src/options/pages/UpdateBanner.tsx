/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useCallback } from "react";
import { Button } from "react-bootstrap";
import "@/layout/Banner";
import { browser } from "webextension-polyfill-ts";
import { useAsyncState } from "@/hooks/common";
import { getAvailableVersion } from "@/background/installer";
import { reportError } from "@/telemetry/logging";

const UpdateBanner: React.FunctionComponent = () => {
  const [updateAvailable] = useAsyncState(async () => {
    try {
      const { installed, available } = await getAvailableVersion();
      return available && installed !== available;
    } catch (error: unknown) {
      reportError(error);
      return false;
    }
  }, []);

  const update = useCallback(() => {
    browser.runtime.reload();
  }, []);

  if (!updateAvailable) {
    return null;
  }

  return (
    <div className="update-banner w-100">
      <div className="mx-auto d-flex">
        <div className="flex-grow-1" />
        <div className="align-self-center">
          An update to PixieBrix is available
        </div>
        <div className="ml-3">
          <Button className="info" size="sm" onClick={update}>
            Update
          </Button>
        </div>
        <div className="flex-grow-1" />
      </div>
    </div>
  );
};

export default UpdateBanner;
