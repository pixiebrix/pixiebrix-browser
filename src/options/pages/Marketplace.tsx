import React, { useCallback, useMemo } from "react";
import { connect } from "react-redux";
import Row from "react-bootstrap/Row";
import Card from "react-bootstrap/Card";
import Button from "react-bootstrap/Button";
import { useToasts } from "react-toast-notifications";
import { useFetch } from "@/hooks/fetch";
import { GridLoader } from "react-spinners";
import { optionsSlice, OptionsState } from "../slices";
import { PageTitle } from "@/layout/Page";
import Rollbar from "rollbar";
import { ensureRecipePermissions } from "@/permissions";
import sortBy from "lodash/sortBy";
import { faStoreAlt } from "@fortawesome/free-solid-svg-icons";
import { Metadata } from "@/core";
import { RecipeDefinition } from "@/types/definitions";

import "./Marketplace.scss";

const { installRecipe } = optionsSlice.actions;

type Install = ({ recipe }: { recipe: RecipeDefinition }) => void;

interface RecipeProps {
  metadata: Metadata;
  installed: boolean;
  onInstall: () => void;
}

const Recipe: React.FunctionComponent<RecipeProps> = ({
  metadata: { name, description, author },
  onInstall,
  installed,
}) => {
  return (
    <div className="col-md-4 col-lg-3 col-sm-2">
      <Card style={{ width: "19rem", minWidth: "19em" }} className="mb-4">
        <Card.Header>{name}</Card.Header>
        <Card.Body>
          <Card.Text>{description}</Card.Text>
          {!installed ? (
            <Button variant="primary" onClick={onInstall}>
              Install
            </Button>
          ) : (
            <Button variant="primary" disabled>
              Installed
            </Button>
          )}
        </Card.Body>
      </Card>
    </div>
  );
};

function useInstall(
  installRecipe: Install
): (recipe: RecipeDefinition) => Promise<void> {
  const { addToast } = useToasts();
  return useCallback(
    async (recipe) => {
      try {
        if (!(await ensureRecipePermissions(recipe))) {
          addToast(
            `You must accept permissions for ${recipe.metadata.name} to install it`,
            {
              appearance: "error",
              autoDismiss: true,
            }
          );
          return;
        }
        installRecipe({ recipe });
        addToast(`Installed ${recipe.metadata.name}`, {
          appearance: "success",
          autoDismiss: true,
        });
      } catch (ex) {
        // @ts-ignore: need to figure out how to import rollbar correctly in typescript
        Rollbar.error(ex);
        console.error(`Error installing ${recipe.metadata.name}`, ex);
        addToast(`Error installing ${recipe.metadata.name}`, {
          appearance: "error",
          autoDismiss: true,
        });
      }
    },
    [installRecipe, addToast]
  );
}

interface MarketplaceProps {
  installedRecipes: Set<string>;
  installRecipe: Install;
}

const Marketplace: React.FunctionComponent<MarketplaceProps> = ({
  installRecipe,
  installedRecipes,
}) => {
  const rawRecipes = useFetch("/api/recipes/") as RecipeDefinition[];

  const recipes = useMemo(
    () =>
      rawRecipes ? sortBy(rawRecipes, (x) => x.metadata.name) : rawRecipes,
    [rawRecipes]
  );

  const install = useInstall(installRecipe);

  return (
    <div className="marketplace-component">
      <PageTitle icon={faStoreAlt} title="Marketplace" />
      <div className="pb-4">
        <p>
          Find and install pre-made blueprints for your favorite websites and
          SaaS apps
        </p>
      </div>

      <Row>
        {(recipes ?? []).map((x) => (
          <Recipe
            key={x.metadata.id}
            {...x}
            onInstall={() => install(x)}
            installed={installedRecipes.has(x.metadata.id)}
          />
        ))}
        {recipes == null && <GridLoader />}
      </Row>
    </div>
  );
};

export default connect(
  ({ options }: { options: OptionsState }) => ({
    installedRecipes: new Set(
      Object.values(options.extensions).flatMap((extensionPoint) =>
        Object.values(extensionPoint)
          .map((x) => x._recipeId)
          .filter((x) => x)
      )
    ),
  }),
  { installRecipe }
)(Marketplace);
