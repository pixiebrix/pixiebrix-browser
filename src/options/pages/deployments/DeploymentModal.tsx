/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useCallback } from "react";
import { DeploymentState } from "@/hooks/useDeployments";
import AsyncButton from "@/components/AsyncButton";
import { DropdownButton, Dropdown, Modal } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { SettingsState } from "@/store/settingsTypes";
import settingsSlice from "@/store/settingsSlice";
import useNotifications from "@/hooks/useNotifications";
import { useUpdateAvailable } from "@/options/pages/UpdateBanner";

const FIFTEEN_MINUTES_MILLIS = 900_000;
const ONE_HOUR_MILLIS = FIFTEEN_MINUTES_MILLIS * 4;
const ONE_DAY_MILLIS = ONE_HOUR_MILLIS * 24;

const SnoozeButton: React.FC<{ snooze: (durationMillis: number) => void }> = ({
  snooze,
}) => (
  <DropdownButton variant="info" id="dropdown-snooze" title="Remind Me Later">
    <Dropdown.Item
      onClick={() => {
        snooze(FIFTEEN_MINUTES_MILLIS);
      }}
    >
      15 Minutes
    </Dropdown.Item>
    <Dropdown.Item
      onClick={() => {
        snooze(ONE_HOUR_MILLIS);
      }}
    >
      1 Hour
    </Dropdown.Item>
    <Dropdown.Item
      onClick={() => {
        snooze(ONE_DAY_MILLIS);
      }}
    >
      1 Day
    </Dropdown.Item>
    <Dropdown.Item
      onClick={() => {
        snooze(ONE_DAY_MILLIS * 3);
      }}
    >
      3 Days
    </Dropdown.Item>
  </DropdownButton>
);

/**
 * Modal to update the browser extension and/or deployments. Can be snoozed
 * @see DeploymentBanner
 * @see UpdateBanner
 */
const DeploymentModal: React.FC<
  Pick<
    DeploymentState,
    "extensionUpdateRequired" | "update" | "updateExtension"
  >
> = ({ extensionUpdateRequired, updateExtension, update }) => {
  const dispatch = useDispatch();
  const notify = useNotifications();
  const hasUpdatesAvailable = useUpdateAvailable();

  const nextUpdate = useSelector<{ settings: SettingsState }>(
    (state) => state.settings.nextUpdate
  );

  const snooze = useCallback(
    (durationMillis: number) => {
      notify.success("Snoozed extensions and deployment updates", {
        event: "SnoozeUpdates",
      });
      dispatch(settingsSlice.actions.snoozeUpdates({ durationMillis }));
    },
    [notify, dispatch]
  );

  if (nextUpdate && nextUpdate > Date.now()) {
    // Snoozed
    return null;
  }

  if (hasUpdatesAvailable) {
    return (
      <Modal show>
        <Modal.Header>
          <Modal.Title>Update Available</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          An update to the PixieBrix browser extension is available. After
          updating, you will need need to reload any pages where PixieBrix is
          running.
        </Modal.Body>
        <Modal.Footer>
          <SnoozeButton snooze={snooze} />

          <AsyncButton variant="primary" onClick={updateExtension}>
            Update
          </AsyncButton>
        </Modal.Footer>
      </Modal>
    );
  }

  if (extensionUpdateRequired) {
    return (
      <Modal show>
        <Modal.Header>
          <Modal.Title>Update Available</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          Update the PixieBrix Browser Extension to activate team deployments.
          After updating, you will need need to reload any pages where PixieBrix
          is running
        </Modal.Body>
        <Modal.Footer>
          <SnoozeButton snooze={snooze} />

          <AsyncButton variant="primary" onClick={updateExtension}>
            Update
          </AsyncButton>
        </Modal.Footer>
      </Modal>
    );
  }

  return (
    <Modal show>
      <Modal.Header>
        <Modal.Title>Team Deployments Available</Modal.Title>
      </Modal.Header>
      <Modal.Body>Team deployments are ready to activate</Modal.Body>
      <Modal.Footer>
        <SnoozeButton snooze={snooze} />
        <AsyncButton variant="primary" onClick={update}>
          Activate
        </AsyncButton>
      </Modal.Footer>
    </Modal>
  );
};

export default DeploymentModal;
