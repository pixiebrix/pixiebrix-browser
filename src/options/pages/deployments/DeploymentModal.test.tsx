/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import DeploymentModal from "@/options/pages/deployments/DeploymentModal";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import settingsSlice, { initialSettingsState } from "@/store/settingsSlice";
import MockDate from "mockdate";
import { SettingsState } from "@/store/settingsTypes";
import { authSlice } from "@/auth/authSlice";
import { useUpdateAvailable } from "@/options/pages/UpdateBanner";
import { AuthState } from "@/auth/authTypes";

jest.mock("@/options/pages/UpdateBanner", () => ({
  useUpdateAvailable: jest.fn(),
}));

browser.runtime.reload = jest.fn();

const reloadMock = browser.runtime.reload as jest.Mock;
const useUpdateAvailableMock = useUpdateAvailable as jest.Mock;

beforeEach(() => {
  reloadMock.mockReset();
});

const renderModal = (
  { extensionUpdateRequired }: { extensionUpdateRequired: boolean },
  settings: Partial<SettingsState>,
  auth: Partial<AuthState>
) => {
  const store = configureStore({
    reducer: {
      settings: settingsSlice.reducer,
      auth: authSlice.reducer,
    },
    preloadedState: {
      settings: { ...settingsSlice.getInitialState(), ...settings },
      auth: { ...authSlice.getInitialState(), ...auth },
    },
  });

  return render(
    <Provider store={store}>
      <DeploymentModal
        update={jest.fn()}
        extensionUpdateRequired={extensionUpdateRequired}
        updateExtension={jest.fn()}
      />
    </Provider>
  );
};

describe("DeploymentModal", () => {
  it("should not render when snoozed", async () => {
    // Tonight I'm going to party like it's 1999
    const date = new Date("12/31/1998");
    MockDate.set(date);

    (useUpdateAvailable as jest.Mock).mockReturnValue(true);

    renderModal(
      {
        extensionUpdateRequired: false,
      },
      {
        ...initialSettingsState,
        nextUpdate: date.getTime() + 1,
      },
      {}
    );

    expect(screen.queryAllByRole("dialog")).toHaveLength(0);
  });

  it("should render if snoozed if not shown to user yet", async () => {
    // Tonight I'm going to party like it's 1999
    const date = new Date("12/31/1998");
    MockDate.set(date);

    (useUpdateAvailable as jest.Mock).mockReturnValue(true);

    renderModal(
      {
        extensionUpdateRequired: false,
      },
      {
        ...initialSettingsState,
        nextUpdate: date.getTime() + 1,
        updatePromptTimestamp: null,
      },
      {
        enforceUpdateMillis: 1,
      }
    );

    expect(screen.queryAllByRole("dialog")).toHaveLength(1);
  });

  it("should render modal when snooze expired", async () => {
    const nextUpdate = new Date("12/31/1998").getTime();
    MockDate.set(new Date(nextUpdate + 1));

    useUpdateAvailableMock.mockReturnValue(true);

    renderModal(
      {
        extensionUpdateRequired: false,
      },
      {
        ...initialSettingsState,
        nextUpdate,
      },
      {}
    );

    expect(await screen.findAllByRole("dialog")).toMatchSnapshot();
    expect(
      screen.queryAllByText(
        "An update to the PixieBrix browser extension is available"
      )
    ).not.toBeNull();
    expect(screen.getByText("Remind Me Later")).not.toBeDisabled();
  });

  it("should render when is enforced", async () => {
    // Tonight I'm going to party like it's 1999
    const time = new Date("12/31/1998").getTime();
    MockDate.set(time);

    (useUpdateAvailable as jest.Mock).mockReturnValue(true);

    renderModal(
      {
        extensionUpdateRequired: false,
      },
      {
        ...initialSettingsState,
        // It is snoozed
        nextUpdate: time + 1,
        // But enforcement window is now enforced
        updatePromptTimestamp: time - 3,
      },
      { enforceUpdateMillis: 1 }
    );

    expect(reloadMock).toHaveBeenCalledTimes(1);
    expect(await screen.findAllByRole("dialog")).toMatchSnapshot();
    expect(screen.getByText("Remind Me Later")).toBeDisabled();
  });

  it("should render deployment update if update is enforced", async () => {
    // Tonight I'm going to party like it's 1999
    const date = new Date("12/31/1998");
    MockDate.set(date);

    (useUpdateAvailable as jest.Mock).mockReturnValue(false);

    renderModal(
      {
        extensionUpdateRequired: false,
      },
      {
        ...initialSettingsState,
        // It is snoozed
        nextUpdate: date.getTime() + 1,
        updatePromptTimestamp: new Date(date.getTime() - 2).getTime(),
      },
      { enforceUpdateMillis: 1 }
    );

    expect(reloadMock).toHaveBeenCalledTimes(0);
    expect(await screen.findAllByRole("dialog")).toMatchSnapshot();
    expect(screen.getByText("Remind Me Later")).toBeDisabled();
  });

  it("should reload browser extension if extension update required for deployment", async () => {
    // Tonight I'm going to party like it's 1999
    const date = new Date("12/31/1998");
    MockDate.set(date);

    (useUpdateAvailable as jest.Mock).mockReturnValue(false);

    renderModal(
      {
        extensionUpdateRequired: true,
      },
      {
        ...initialSettingsState,
        // It is snoozed
        nextUpdate: date.getTime() + 1,
        updatePromptTimestamp: new Date(date.getTime() - 2).getTime(),
      },
      { enforceUpdateMillis: 1 }
    );

    expect(reloadMock).toHaveBeenCalledTimes(1);
    expect(await screen.findAllByRole("dialog")).toMatchSnapshot();
    expect(screen.getByText("Remind Me Later")).toBeDisabled();
  });
});
