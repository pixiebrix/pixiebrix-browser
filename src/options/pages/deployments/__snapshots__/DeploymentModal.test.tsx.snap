// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentModal should reload browser extension if extension update required for deployment 1`] = `
Array [
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Update Available
          </div>
        </div>
        <div
          class="modal-body"
        >
          Update the PixieBrix Browser Extension to activate team deployments. After updating, you will need need to reload any pages where PixieBrix is running
        </div>
        <div
          class="modal-body"
        >
          <div
            class="fade alert alert-info show"
            role="alert"
          >
            Your team admin has set a policy requiring you to apply updates.
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <div
            aria-disabled="true"
            class="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-info"
              disabled=""
              id="dropdown-snooze"
              type="button"
            >
              Remind Me Later
            </button>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Update
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`DeploymentModal should render deployment update if update is enforced 1`] = `
Array [
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Team Deployments Available
          </div>
        </div>
        <div
          class="modal-body"
        >
          Team deployments are ready to activate
        </div>
        <div
          class="modal-body"
        >
          <div
            class="fade alert alert-info show"
            role="alert"
          >
            Your team admin has set a policy requiring you to apply updates.
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <div
            aria-disabled="true"
            class="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-info"
              disabled=""
              id="dropdown-snooze"
              type="button"
            >
              Remind Me Later
            </button>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Activate
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`DeploymentModal should render modal when snooze expired 1`] = `
Array [
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Update Available
          </div>
        </div>
        <div
          class="modal-body"
        >
          An update to the PixieBrix browser extension is available. After updating, you will need need to reload any pages where PixieBrix is running.
        </div>
        <div
          class="modal-footer"
        >
          <div
            aria-disabled="false"
            class="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-info"
              id="dropdown-snooze"
              type="button"
            >
              Remind Me Later
            </button>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Update
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`DeploymentModal should render when is enforced 1`] = `
Array [
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Update Available
          </div>
        </div>
        <div
          class="modal-body"
        >
          An update to the PixieBrix browser extension is available. After updating, you will need need to reload any pages where PixieBrix is running.
        </div>
        <div
          class="modal-body"
        >
          <div
            class="fade alert alert-info show"
            role="alert"
          >
            Your team admin has set a policy requiring you to apply updates.
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <div
            aria-disabled="true"
            class="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-info"
              disabled=""
              id="dropdown-snooze"
              type="button"
            >
              Remind Me Later
            </button>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Update
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
